### PR TITLE
initial_commit to pysteps-nwp-importers

### PR DESCRIPTION
**Pull request by RubenImhoff**
This PR is a first step to a separate nwp importers plugin, which can be used in the pysteps repo. It resolves issue #218 in the pysteps repo and was discussed in the `blending-v1` branch in the pysteps repo.

To do:

- [x] Make sure tests run automatically
- [x] Check if tests run and adjust (I had some problems with the BoM NWP importer via the plugin)
- [x] Download files from `pysteps-data-master`, needed for tests

### CODE REVIEW

**Comment by RubenImhoff** [comment]
Alright, from here I might need some help. As far as I understand CI, I have tried to make some commits, but not sure if this is what it should be. 

Two other points of attention:

- BoM importer and test fail, but I haven't figured out why. Perhaps it is related to `@postprocess_import()`. 
- Right now all test data is downloaded, but in principle we only need the nwp folder.

**Comment by aperezhortal** [comment]
> BoM importer and test fail, but I haven't figured out why. Perhaps it is related to @postprocess_import().

You are right, there was a problem with the cookie-cutter template :sweat_smile:  (I will fix it today). The postprocess_import decorator is not needed since it is being applied during the plugin discovery. 

Commit 85e5085 removes the decorator, and also other imports from pysteps to avoid any potential issue with circular dependencies.

I will try to take a look at the CI workflow today.

**Comment by RubenImhoff** [comment]
Wonderful! I only still keep getting an error when trying to import the BoM data. Let's see if you have the same, it might be on my side, it gives a `Windows fatal exception: access violation`.

**Comment by aperezhortal** [comment]
The tests are up and running :rocket: 

But....there is a problem with windows and py3.9 that seems related to mamba. For some reason, it is trying to install pysteps=1.0.0...

**Comment by RubenImhoff** [comment]
Great work, @aperezhortal! Yes, looks like it has trouble installing mamba - no idea why.. 
Good to see the other tests pass! Locally they do too, except for the BoM importer, which oddly enough still returns a `RuntimeError: NetCDF: HDF error` on my Python 3.9 installation.

**Comment by dnerini** [comment]
Great work @RubenImhoff and @aperezhortal! 

Going forward we should think whether it would make more sense to separate the importers into individual plugins, instead of having them all toghether. Moreover, ideally the importer plugins will be maintained directly by the data owners (ie. BOM, KNMI, RMI) and hosted on their Github page, but let's make one step at the time ;-)

More importantly: currently the only way to install the plugin is with the source code, but I don't see why we shouldn't publish it on pypi? This will also facilitate our CI workflows with the pysteps base package, since it will only take one `pip install` to get the importer. It should be rather easy to implement the relevant github workflow and update the documentation. If you agree, I'll make the needed changes.

File: docs/installation.rst [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+.. The following installation instructions are recommended if the plugin needs to be installed
+   from its sources (e.g. directly from the github repo).
+
+From sources
+------------
+
+The latest development version of pysteps_nwp_importers can be installed using
+pip by running in a terminal::
+
+    pip install git+https://github.com/YOUR_GITHUB_USERNAME/pysteps_nwp_importers
```
Comment (dnerini) [inline docs/installation.rst]:
```suggestion
    pip install git+https://github.com/pySTEPS/pysteps-nwp-importers
```

File: docs/installation.rst [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+============
+Installation
+============
+
+.. The following installation instructions are recommended if the plugin is uploaded to pypi.
+
+Stable release
+--------------
+
+To install pysteps_nwp_importers, run this command in your terminal:
```
Comment (dnerini) [inline docs/installation.rst]: Isn't the name of the package "pysteps-nwp-importers"?

File: docs/installation.rst [hunk 3]
```diff
@@ -0,0 +1,40 @@
+.. highlight:: shell
+
+.. IMPORTANT: Update this file with your recommended installation instructions,
+   As an example, two different installation instructions are shown next.
+
+============
+Installation
+============
+
+.. The following installation instructions are recommended if the plugin is uploaded to pypi.
```
Comment (dnerini) [inline docs/installation.rst]: Let's keep this only if the plugins if published on pypi

File: docs/requirements.txt [hunk 4]
```diff
@@ -0,0 +1,4 @@
+# Additional requeriments related to the documentation build only
```
Comment (dnerini) [inline docs/requirements.txt]:
```suggestion
# Additional requirements related to the documentation build only
```

File: requirements.txt [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+pip
+wheel
+watchdog
+tox
+coverage
+Sphinx
+numpy
+xarray
+netCDF4
+pysteps
```
Comment (dnerini) [inline requirements.txt]:
```suggestion
pysteps>=1.4,<2.0
```

I suggest to explicitly state the minimum (1.4 introduced the plugins) and maximum (2.0 will likely change the interface to the importers) versions here.

File: requirements_dev.txt [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+watchdog
+tox
+coverage
+Sphinx
+pytest
+pytest-runner
+numpy
+xarray
+netCDF4
+pysteps
```
Comment (dnerini) [inline requirements_dev.txt]:
```suggestion
pysteps>=1.4,<2.0
```

(see above)

File: setup.py [hunk 7]
```diff
@@ -0,0 +1,42 @@
+from setuptools import setup, find_packages
+
+with open("README.rst") as readme_file:
+    readme = readme_file.read()
+
+requirements = ["numpy", "xarray", "dask", "netCDF4"]
```
Comment (dnerini) [inline setup.py]:
```suggestion
requirements = ["numpy", "xarray", "dask", "netCDF4", "pysteps>=1.4,<2.0"]
```

File: setup.py [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+
+setup(
+    author="Ruben Imhoff",
+    author_email="ruben.imhoff@deltares.nl",
+    python_requires=">=3.7",
+    classifiers=[
+        "License :: OSI Approved :: BSD License",
+        "Programming Language :: Python :: 3",
+        "Programming Language :: Python :: 3.7",
+        "Programming Language :: Python :: 3.8",
```
Comment (dnerini) [inline setup.py:19]:
```suggestion
        "Programming Language :: Python :: 3.8",
        "Programming Language :: Python :: 3.9",        
```

File: setup.py [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+    packages=find_packages(),
+    setup_requires=setup_requirements,
+    entry_points={
+        "pysteps.plugins.importers": [
+            "import_knmi_nwp=pysteps_nwp_importers.importer_knmi_nwp:import_knmi_nwp",
+            "import_bom_nwp=pysteps_nwp_importers.importer_bom_nwp:import_bom_nwp",
+            "import_rmi_nwp=pysteps_nwp_importers.importer_rmi_nwp:import_rmi_nwp",
+        ]
+    },
+    version="1.6",
```
Comment (dnerini) [inline setup.py]: I don't think this should refer to the version from the base package, but rather use its own. Should we go for 0.1.0?

File: ci/ci_test_env.yml [hunk 10]
```diff
@@ -0,0 +1,15 @@
...
+channels:
+  - conda-forge
+  - defaults
+dependencies:
+  - python>=3.7
+  - pip
+  - mamba
+  - numpy
+  - netCDF4
+  - pysteps
```
Comment (dnerini) [inline ci/ci_test_env.yml]:
```suggestion
  - pysteps>=1.4,<2.0
```

File: setup.py (hunk 9 above)
Comment (RubenImhoff) [reply setup.py]: Good point, I agree!

File: docs/installation.rst (hunk 2 above)
Comment (RubenImhoff) [reply docs/installation.rst]: Correct! I started of with pysteps_nwp_importers and then changed it to be more consistent with the other repositories. I'll change it.

**Comment by RubenImhoff** [comment]
> Going forward we should think whether it would make more sense to separate the importers into individual plugins, instead of having them all toghether. Moreover, ideally the importer plugins will be maintained directly by the data owners (ie. BOM, KNMI, RMI) and hosted on their Github page, but let's make one step at the time ;-)

Hmm yeah, it would make sense to do it separately, but I can imagine the pySTEPS base repo filling up quickly with different importers if we don't directly move it to local Github pages. Should we just continue like this for now and keep this as a next step? (no strong opinion here, whatever works!)

> More importantly: currently the only way to install the plugin is with the source code, but I don't see why we shouldn't publish it on pypi? This will also facilitate our CI workflows with the pysteps base package, since it will only take one `pip install` to get the importer. It should be rather easy to implement the relevant github workflow and update the documentation. If you agree, I'll make the needed changes.

Good idea!

**Comment by dnerini** [comment]
Mmmmh the python 3.9 windows test keeps failing. A bit weird since the same setup seems to work fine in https://github.com/pySTEPS/pysteps/pull/257 ...

Not sure what we could try next, maybe using conda to setup mamba as in https://github.com/pySTEPS/pysteps/pull/256 ?

**Comment by RubenImhoff** [comment]
It is a bit weird, indeed.. 

> Not sure what we could try next, maybe using conda to setup mamba as in pySTEPS/pysteps#256 ?

Yeah, why not, let's try that!

**Comment by dnerini** [comment]
The tests on windows pass when pysteps is installed with pip, which could be an indication that something's wrong with the relative conda-forge package (did we remove windows-py39 from conda-forge? @aperezhortal ?).

Unforuntaly using pip now causes the (usual) issue with gcc on macOS, which I fixed with a not-so-nice workaround (preinstall step, hardcode version min-max). 

Not super elegant, but good enough for me at this point, you agree?

**Review by dnerini — APPROVED** [review]

**Comment by RubenImhoff** [comment]
Great work, @dnerini! Ready to merge?

**Comment by aperezhortal** [comment]
@dnerini, we are not building the conda package for python3.9 in windows: https://github.com/conda-forge/pysteps-feedstock/blob/master/.azure-pipelines/azure-pipelines-win.yml 

But I don't remember why. Probably we remove it because some dependencies were missing back then.

We can merge this PR and later on update the workflow when the py39 package is built for windows.

**Comment by aperezhortal** [comment]
> Going forward we should think whether it would make more sense to separate the importers into individual plugins, instead of having them all together. Moreover, ideally, the importer plugins will be maintained directly by the data owners (ie. BOM, KNMI, RMI) and hosted on their Github page, but let's make one step at the time ;-)

I agree with the idea of transferring the maintenance responsibilities to the data owners since custom importers are kind of outside of the scope of pysteps. 

We can move forward with this PR to have a functional example. But before would not upload it to pypi at this moment if the importer will be maintained by another organization or individual. They can upload it to pypi.

**Comment by dnerini** [comment]
> We can move forward with this PR to have a functional example. But before would not upload it to pypi at this moment if the importer will be maintained by another organization or individual. They can upload it to pypi.

Mh yes, i see your point. Ok let's not publish to pypi for now, meaning that we'll have to install from source for the example gallery (which is totally ok)

**Comment by dnerini** [comment]
update: with d8848f0f53b97b3a9d3bb39138d56950a5d53af2 we removed the option to install from pypi.

**Pull request merged**